File: cpp/chapter5/5-2-1.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <tuple>
#include <vector>
using namespace std;

template <typename CapacityTp, typename CostTp> class MinCostFlow {
  public:
    using Edge = tuple<int, CapacityTp, CostTp, int>;

    MinCostFlow(int V) {
        V_ = V;
        inf_cost_ = numeric_limits<CostTp>::max();
        G_.resize(V_);
    }

    void add_edge(int u, int v, CapacityTp capacity, CostTp cost) {
        G_[u].emplace_back(v, capacity, cost, G_[v].size());
        G_[v].emplace_back(u, 0, -cost, G_[u].size() - 1);
    }

    tuple<vector<CostTp>, vector<int>, vector<int>> bellman_ford(int s) {
        vector<CostTp> dist(V_, inf_cost_);
        dist[s] = 0;
        vector<int> pv(V_, 0);
        vector<int> pe(V_, 0);
        while(true) {
            bool update = false;
            for(int v = 0; v < V_; v++) {
                if(dist[v] == inf_cost_) {
                    continue;
                }
                for(int i = 0; i < (int)G_[v].size(); i++) {
                    auto [nxt, capacity, cost, rev] = G_[v][i];
                    if(capacity > 0 && dist[nxt] > dist[v] + cost) {
                        dist[nxt] = dist[v] + cost;
                        update = true;
                        pv[nxt] = v;
                        pe[nxt] = i;
                    }
                }
            }
            if(!update) {
                break;
            }
        }
        return make_tuple(dist, pv, pe);
    }

    CostTp calc_min_cost_flow(int s, int t, CapacityTp f) {
        CostTp result = 0;
        while(f > 0) {
            auto [dist, pv, pe] = bellman_ford(s);
            if(dist[t] == inf_cost_) {
                return inf_cost_;
            }
            CapacityTp flow = f;
            int v = t;
            while(v != s) {
                flow = min(flow, get<1>(G_[pv[v]][pe[v]]));
                v = pv[v];
            }
            result += flow * dist[t];
            f -= flow;
            v = t;
            while(v != s) {
                // 辺の容量を更新する
                get<1>(G_[pv[v]][pe[v]]) -= flow;
                // 逆辺についても同様にする
                int rev = get<3>(G_[pv[v]][pe[v]]);
                get<1>(G_[v][rev]) += flow;
                v = pv[v];
            }
        }
        return result;
    }

  private:
    int V_;
    vector<vector<Edge>> G_;
    CostTp inf_cost_;
};

int main() {
    int V, E, F;
    cin >> V >> E >> F;
    MinCostFlow<long long, long long> mincost_flow(V);
    for(int i = 0; i < E; i++) {
        int u, v, capacity, cost;
        cin >> u >> v >> capacity >> cost;
        u--, v--;
        mincost_flow.add_edge(u, v, capacity, cost);
    }
    cout << mincost_flow.calc_min_cost_flow(0, V - 1, F) << endl;
    return 0;
}

File: python/chapter5/5-2-1.py
INF = 1e100
V, E, F = map(int, input().split())
G = [[] for _ in range(V)]


def add_edge(u, v, capacity, cost):
    G[u].append((v, capacity, cost, len(G[v])))
    G[v].append((u, 0, -cost, len(G[u]) - 1))


def bellman_ford(s):
    dist = [INF] * V
    dist[s] = 0
    pv = [0] * V
    pe = [0] * V
    while True:
        update = False
        for v in range(V):
            if dist[v] == INF:
                continue
            for i in range(len(G[v])):
                next, capacity, cost, _ = G[v][i]
                if capacity > 0 and dist[next] > dist[v] + cost:
                    dist[next] = dist[v] + cost
                    update = True
                    pv[next] = v
                    pe[next] = i
        if not update:
            break
    return dist, pv, pe


def calc_min_cost_flow(s, t, f):
    result = 0
    while f > 0:
        dist, pv, pe = bellman_ford(s)
        if dist[t] == INF:
            return INF
        flow = f
        v = t
        while v != s:
            flow = min(flow, G[pv[v]][pe[v]][1])
            v = pv[v]
        result += flow * dist[t]
        f -= flow
        v = t
        while v != s:
            # 辺の情報を持つタプルを分解する
            dest, capacity, cost, r = G[pv[v]][pe[v]]
            # 辺の容量を更新する
            capacity -= flow
            # 書き換え終わったので、タプルに戻す
            G[pv[v]][pe[v]] = (dest, capacity, cost, r)
            # 逆辺についても同様にする
            rev = G[pv[v]][pe[v]][3]
            dest, capacity, cost, r = G[v][rev]
            capacity += flow
            G[v][rev] = (dest, capacity, cost, r)
            v = pv[v]
    return result


for _ in range(E):
    u, v, capacity, cost = map(int, input().split())
    u -= 1
    v -= 1
    add_edge(u, v, capacity, cost)
ans = calc_min_cost_flow(0, V - 1, F)
print(ans)
